packages/storage/src/indexed-db/stream.ts

```diff
@@ -0,0 +1,24 @@
+import { AnyMessage, JsonValue, Message, MessageType } from '@bufbuild/protobuf';
+import { PenumbraDb, PenumbraStoreNames } from '@penumbra-zone/types';
+import { IDBPCursorWithValue } from 'idb';
+
+export class IdbCursorSource<N extends PenumbraStoreNames, T extends Message<T> = AnyMessage>
+  implements UnderlyingDefaultSource<T>
+{
+  constructor(
+    private cursor: Promise<IDBPCursorWithValue<PenumbraDb, [N], N> | null>,
+    private messageType: MessageType<T>,
+  ) {}
+
+  start(cont: ReadableStreamDefaultController<T>) {
+    // immediately stream as fast as possible, to prevent idb from closing the tx
+    void (async () => {
+      let cursor = await this.cursor;
+      while (cursor) {
+        cont.enqueue(this.messageType.fromJson(cursor.value as JsonValue));
+        cursor = await cursor.continue();
+      }
+      cont.close();
+    })();
+  }
```

Comment:
Ah, clever work around. But does this mean that all of the values are loaded up in the stream (aka in memory) all at once? In the end, is this the equivalent of a getAll()?

Reply:
If this is really the equivalent of `getAll()`, then we are not optimally filtering. In this case we should filter inside the `while (cursor)` loop to avoid reading all data.